src/lib.rs

```diff
@@ -507,41 +560,90 @@ impl IntoRawFd for CANSocket {
 
 impl Drop for CANSocket {
     fn drop(&mut self) {
-        self.close().ok(); // ignore result
+        self.close()
+            // .map_err(|e|eprintln!("CANSocket close error: {}", e.to_string()))
+            .ok(); // ignore result
     }
 }
 
-/// CANFrame
+/// CANFrameStruct
 ///
 /// Uses the same memory layout as the underlying kernel struct for performance
 /// reasons.
 #[derive(Debug, Copy, Clone)]
 #[repr(C)]
-pub struct CANFrame {
+struct CANFrameStruct {
     /// 32 bit CAN_ID + EFF/RTR/ERR flags
     _id: u32,
 
-    /// data length. Bytes beyond are not valid
+    /// frame payload length in bytes (0 .. 8) for classical frame,
+    /// (0 .. 64) for FD frame
     _data_len: u8,
 
-    /// padding
-    _pad: u8,
+    /// padding for classical frame, additional
+    /// flags for CAN FD
+    _pad_flags: u8,
 
     /// reserved
     _res0: u8,
 
     /// reserved
     _res1: u8,
 
-    /// buffer for data
-    _data: [u8; 8],
+    /// buffer for data (classical or FD frame)
+    _data: [u8; CAN_FRAME_DATA_LEN_MAX]
+}
+
+enum CANFrameType {
+    Normal,
+
+    #[cfg(feature = "can_fd")]
+    Fd
+}
+
+pub struct CANFrame {
+    tag: CANFrameType,
```

Comment:
I am not very familiar with CanFD,  but it looks like the header of a frame contains different information compared to a classical CAN frame. E.g. no RTR flag, but adds EDL, ERS and BRS. 

From an API perspective wouldn't it make more sense to have a completely separate CanFdFrame struct. Then you can also easily implement functions for EDL, ERS, and BRS and drop functions e.g. for setting RTR. You'd need a separate `read_fd_frame then but i think it would make using it easier.

Reply:
On the bus level there is EDL bit indicating a FD frame, but this bit is not present in SocketCAN data structures. In fact, the only way one can determine if a frame is classical or FD is from the size returned by `read` (or `recvmsg`) syscall (please read SocketCAN documentation: https://www.kernel.org/doc/Documentation/networking/can.txt - Remark about CAN FD (flexible data rate) support).

That SocketCAN API is also the reason why you can't have separate struct and function for reading CAN FD - you don't know whether you get classical or FD frame until you read it.